### ISSUE TITLE
Missing ConfigureAwait(false) in some parts of code

### ISSUE THREAD

**Issue by pecanw** [quartznet/quartznet#618]
### Version: x.y.z
3.0.4

### Expected behavior
ConfigureAwait(false) is used in all library code, i.e. no deadlock when calling some asynchronous operations from synchronous function.

### Actual behavior
ConfigureAwait(false) is not used and a deadlock can occur.

### Steps to reproduce
e.g. calling `IJobStore.RetrieveJob(jobKey).ConfigureAwait(false).GetAwaiter().GetResult()` on an AdoJobStore from a non-async controller action

### Fix:
[Pull request #617](https://github.com/quartznet/quartznet/pull/617) created.

I haven't included unit tests for this issue as it is quite difficult to write it in this case (applying following rule from guidelines) 
> If there is a scenario that is far too hard to test there does not need to be a test for it.

**Comment by pecanw** [comment]
I would really appreciate merging the fix to the master repo and releasing an official nuget with the fix. Can  an owner of the repo have a look please?

**Comment by lahma** [comment]
Thanks for the report and the fix. Sorry it took such a long time to process. Hopefully can do a release the upcoming weekend.

**Comment by pecanw** [comment]
Great news! Thanks.

**Comment by AndersMad** [comment]
Won't this potentially break `await *` usage in ASP.NET? As `ConfigureAwait(false)` will not return to the current `HttpContext`.

**Comment by lahma** [comment]
Generally you can't access contextual items anyway as jobs etc run on different threads. Quartz tries to ensure that no (outside) context is catched as part of processing.

**Comment by AndersMad** [comment]
I'm thinking about calls that e.g. gets info about running jobs from within the ASP.NET app - these calls are also async.

**Comment by lahma** [comment]
Those running jobs would still be inside a singleton (RAMJobStore) that does not have 'current' web context.  Maybe a failing example would be easier to understand and think workarounds for.

**Comment by AndersMad** [comment]
Sry, never mind: I believe that it doesn't matter if `ConfigureAwait(false)` is used - as long as the caller to the API doesn't have it - it should restore the context at that call.

**Comment by pecanw** [comment]
I recommend to read this article - [Async/Await - Best Practices in Asynchronous Programming](https://msdn.microsoft.com/en-us/magazine/jj991977.aspx).

TL;DR: Use `ConfigureAwait(false)` in the library projects and `ConfigureAwait(true)` in the ASP.NET / application. 

You will get back to the original context if you call Quartz following way:
```csharp
public async Task MyActionAsync()
{
    await Quartz.AnyAsyncQuartzFunctionUsingConfigureAwaitFalseInternaly().ConfigureAwait(true);
    // Here you have your context back
}
```

**Comment by AndersMad** [comment]
A bit off topic here, but `ConfigureAwait(true)` is default - so no need for that - and `GetAwaiter().GetResult()` is a bundle of hurt in multi-threaded IO hefty exception throwing code - at least when doing "UI" apps :)

In a later article Stephen says "Using ConfigureAwait(false) to avoid deadlock is at best just a hack)" and links to this https://msdn.microsoft.com/en-us/magazine/mt238404.aspx (2015)

But I'm not arguing against this change - as long as caller doesn't have the `ConfigureAwait(false)` all should be good in UI/sync.context/thread sensitive apps.
